profiling/src/profiling.rs

```diff
@@ -0,0 +1,806 @@
+    pub tags: Vec<Tag>,
+    pub uri: String,
+}
+
+#[derive(Debug)]
+pub struct SampleData {
+    pub frames: Vec<ZendFrame>,
+    pub labels: Vec<Label>,
+    pub sample_values: Vec<i64>,
+}
+
+#[derive(Debug)]
+pub struct SampleMessage {
+    pub key: ProfileIndex,
+    pub value: SampleData,
+}
+
+pub struct Globals {
+    pub interrupt_count: AtomicU32,
+    pub last_interrupt: SystemTime,
+    // todo: current_profile
+}
+
+impl Default for Globals {
+    fn default() -> Self {
+        Self {
+            interrupt_count: AtomicU32::new(0),
+            last_interrupt: SystemTime::now(),
+        }
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct ZendFrame {
+    pub function: Option<String>,
+    pub file: Option<String>,
+    pub line: u32, // use 0 for no line info
+}
+
+unsafe fn zend_string_to_bytes<'a>(ptr: *const zend_string) -> &'a [u8] {
+    if ptr.is_null() {
+        return b"";
+    }
+    let zstr = &*ptr;
+    if zstr.len == 0 {
+        b""
+    } else {
+        std::slice::from_raw_parts(zstr.val.as_ptr() as *const u8, zstr.len as usize)
+    }
+}
+
+/// Extract the "function name" component for the frame. This is a string which
+/// looks like this for methods:
+///     {module}|{class_name}::{method_name}
+/// And this for functions:
+///     {module}|{function_name}
+/// Where the "{module}|" is present only if it's an internal function.
+/// Namespaces are part of the class_name or function_name respectively.
+/// Closures currently repeat the namespace for both class and method, but this
+/// could be changed (should be changed? we don't know yet)
+unsafe fn extract_function_name(func: &zend_function) -> Result<String, Utf8Error> {
+    let method_name: &[u8] = zend_string_to_bytes(func.common.function_name);
+
+    /* The top of the stack seems to reasonably often not have a function, but
+     * still has a scope. I don't know if this intentional, or if it's more of
+     * as situation where scope is only valid if the func is present. So,
+     * erring on the side of caution.
+     */
+    if method_name.is_empty() {
+        return Ok(String::new());
+    }
+
+    // User functions do not have a "module". Maybe one day use composer info?
+    let module: &[u8] = if func.type_ == ZEND_INTERNAL_FUNCTION as u8
+        && !func.internal_function.module.is_null()
+        && !(*func.internal_function.module).name.is_null()
+    {
+        CStr::from_ptr((*func.internal_function.module).name as *const c_char).to_bytes()
+    } else {
+        b""
+    };
+
+    let class_name: &[u8] = if !func.common.scope.is_null() {
+        zend_string_to_bytes((*func.common.scope).name)
+    } else {
+        b""
+    };
+
+    let mut buffer = Vec::<u8>::new();
+    // +1 for "|"
+    let module_len = if module.is_empty() {
+        0
+    } else {
+        module.len() + 1
+    };
+    // +2 for "::"
+    let class_len = if class_name.is_empty() {
+        0
+    } else {
+        class_name.len() + 1
+    };
```

Comment:
Should this be `+ 2`?

Reply:
Since we're not using `try_reserve`, I simplified this whole section so the math errors just can't happen. Thanks!